packages/grid/x-data-grid/src/hooks/features/columnGrouping/useGridColumnGrouping.ts

```diff
@@ -0,0 +1,183 @@
+import * as React from 'react';
+import { GridApiCommunity } from '../../../models/api/gridApiCommunity';
+import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
+import { GridStateInitializer } from '../../utils/useGridInitializeState';
+import {
+  GridColumnGroupingModel,
+  GridColumnNode,
+  isLeaf,
+} from '../../../models/gridColumnGrouping';
+import { gridColumnGroupsLookupSelector } from './gridColumnGroupsSelector';
+import { gridColumnLookupSelector } from '../columns/gridColumnsSelector';
+import { GridColumnGroupLookup } from './gridColumnGroupsInterfaces';
+import { GridColumnGroupingApi } from '../../../models/api/gridColumnGroupingApi';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridStateColDef, GridColDef } from '../../../models/colDef';
+
+export function hasGroupPath(
+  lookupElement: GridColDef | GridStateColDef,
+): lookupElement is GridStateColDef {
+  return (<GridStateColDef>lookupElement).groupPath !== undefined;
+}
+
+type UnwrappedGroupingModel = { [key: string]: string[] };
+
+// This is the recurrence function that help writing `unwrapGroupingColumnModel()`
+const recurrentUnwrapGroupingColumnModel = (
+  columnGroupNode: GridColumnNode,
+  parents: any,
+): UnwrappedGroupingModel => {
+  if (isLeaf(columnGroupNode)) {
+    return { [columnGroupNode.field]: parents };
+  }
+  const rep: UnwrappedGroupingModel = {};
+  const { groupId, children } = columnGroupNode;
+  children.forEach((child) => {
+    const unwrappedSubTree = recurrentUnwrapGroupingColumnModel(child, [...parents, groupId]);
+    Object.entries(unwrappedSubTree).forEach(([key, value]) => {
```

Comment:
As an optimization step, we could avoid this check on each tree level by passing `rep` to `recurrentUnwrapGroupingColumnModel`, then checking for `rep[field]` before mutating it and throwing error if it's already there.
What do you think?